src/postamble.js

```diff
@@ -397,6 +397,19 @@ function exit(status, implicit) {
   Module['quit'](status, new ExitStatus(status));
 }
 
+function getCircularReplacer() {
+  var seen = new WeakSet();
+  return function (key, value) {
+    if (typeof value === "object" && value !== null) {
+      if (seen.has(value)) {
+        return "[Circular]";
+      }
+      seen.add(value);
+    }
+    return value;
+  };
+};
```

Comment:
maybe declare this inside of abort since its only used in there?

What kind of situation would trigger this exactly?  can we create a test for it somehow?

Reply:
Putting it inside abort would be good.

It's triggered when [`process["on"]("unhandledRejection",abort)`](https://github.com/emscripten-core/emscripten/blob/incoming/src/shell.js#L160) is called in nodejs (browser seems to work fine as it doesn't have the event). You can probably test it by calling `process.emit("unhandledRejection", [CircularObject])` manually, after which abort will throw an exception as shown in the original example.

Reply:
I suggest maybe we change that line to `process["on"]("unhandledRejection",onUnhandledRejection)` which can do whatever it wants and then call abort with a string value.

Also we can/should probably just simplify abort to call ".toString()" on its argument.  I'm not sure it makes sense for abort to accept complex objects that need JSON formatting.

Reply:
> Also we can/should probably just simplify abort to call ".toString()" on its argument.

That might also make the cause less obvious, often when calling `toString()` javascript will return `[object Object]`. 

If I remember correctly `process["on"]("unhandledRejection",onUnhandledRejection)` will stop nodejs from showing that there was an unhandledRejection, which means that the only feedback you get (by default) would be from the module.

Reply:
People shouldn't be calling `abort` with arbitrary JS objects.    I think its reasonable to expect it to take a string or an integer.   Anyway, I guess I'm ok with this change as is now.

Reply:
Hmm, actually I agree. This is a little more complexity than it's worth.

I agree we should avoid crashing on a cyclic object, so just removing the JSON.stringify seems like the right thing to do.

Reply:
> People shouldn't be calling abort with arbitrary JS objects.

In the normal case where abort is called from emscripten generated code it wouldn't cause a problem (as far as I know). Changing the `unhandledRejection` to it's own function should also fix the problem, but there might be a lot of overlapping logic from the abort handler.

I do wonder why it catches `unhandledRejection` at all, I haven't look too far into the project structure but I can't imagine a lot of cases where that would actually be applicable.

Reply:
I'm not sure, but I think we caught unhandledRejection because otherwise node didn't show a proper error during async wasm compilation. It's possible it's no longer necessary.

Reply:
That's interesting, well to resolve the issue I see a few options:
1.put the `unhandledRejection` catcher itself behind a flag
2. fix `abort` to be compatible with an external `unhandledRejection` (this PR)
3. remove the `unhandledRejection` (might break something if it is actually used somewhere)

The last one would be the best option in my opinion, but only if we can confirm that it is safe to remove

Reply:
@kripken I found an old issue referring to the reason of catching unhandledRejection,
https://github.com/emscripten-core/emscripten/issues/7855

It seems that there is/was a test case for it, so I might check if it's still valid

Reply:
Thanks, yes, let's see if we still need to have unhandledRejection code. If not, then it would be nice to not have to add code to support circularity etc. as discussed earlier.

Reply:
I've looked at the test and it still fails if you comment `unhandledRejection`, but not for the right reasons. I've changed abort to use process['exit'] (and console.error) instead of throw, and now it does work without `unhandledRejection`.  I'll be running the full test suite to make sure it doesn't break something else. If everything seems reasonable I'll create another PR.